src/Adapter/Carrier/Validate/CarrierValidator.php

```diff
@@ -84,4 +88,39 @@ public function validateLogoUpload(string $filePath): void
             throw new MemoryLimitException('Cannot upload image due to memory restrictions');
         }
     }
+
+    /**
+     * @throws CoreException
+     */
+    private function validateGeneral(Carrier $carrier): void
+    {
+        $this->validateObjectModelProperty($carrier, 'name', CarrierConstraintException::class, CarrierConstraintException::INVALID_NAME);
+        $this->validateObjectModelProperty($carrier, 'grade', CarrierConstraintException::class, CarrierConstraintException::INVALID_GRADE);
+        $this->validateObjectModelProperty($carrier, 'url', CarrierConstraintException::class, CarrierConstraintException::INVALID_TRACKING_URL);
+        $this->validateObjectModelProperty($carrier, 'position', CarrierConstraintException::class, CarrierConstraintException::INVALID_POSITION);
+        $this->validateObjectModelLocalizedProperty($carrier, 'delay', CarrierConstraintException::class, CarrierConstraintException::INVALID_DELAY);
+    }
+
+    /**
+     * @throws CoreException
+     */
+    private function validateShippingGeneral(Carrier $carrier): void
+    {
+        $this->validateObjectModelProperty($carrier, 'shipping_handling', CarrierConstraintException::class, CarrierConstraintException::INVALID_SHIPPING_HANDLING);
+        $this->validateObjectModelProperty($carrier, 'is_free', CarrierConstraintException::class, CarrierConstraintException::INVALID_IS_FREE);
+        $this->validateObjectModelProperty($carrier, 'shipping_method', CarrierConstraintException::class, CarrierConstraintException::INVALID_SHIPPING_METHOD);
+        $this->validateObjectModelProperty($carrier, 'range_behavior', CarrierConstraintException::class, CarrierConstraintException::INVALID_RANGE_BEHAVIOR);
+
+        // A Carrier cannot be both shipping handling and free
+        if ($carrier->shipping_handling && $carrier->is_free) {
```

Comment:
Question, I believe in the UX when free shipping is enabled the other inputs are hidden So it means you can't disable shipping handling once free shipping is enabled

Now the CQRS implementation is not supposed to be influenced by the UX, but it makes me think nonetheless: should this be considered an error or should the `shipping_handling` be disabled automatically when free shipping is enabled? These are two valid approaches so I don't mind either but it's worth thinking about it at least

Reply:
I like this approach that the `shipping_handling` be disabled automaticaly if `is_free` enabled instead of throwing an exception. I will change that in that way.